Kotlin 2.3.20 support

There are breaking changes in `2.3.20-Beta1`:

```
e: org.jetbrains.kotlin.fir.pipeline.IrGenerationExtensionException: 'org.jetbrains.kotlin.ir.declarations.IrDeclarationOriginImpl org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin$Companion.getDEFINED()'
        at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:584)
        at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.runActualizationPipeline(convertToIr.kt:283)
        at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.convertToIrAndActualize(convertToIr.kt:154)
        at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize-MT2kVtw(convertToIr.kt:119)
        at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize-MT2kVtw$default(convertToIr.kt:92)
        at org.jetbrains.kotlin.cli.jvm.compiler.legacy.pipeline.JvmCompilerPipelineKt.convertToIrAndActualizeForJvm-_I1IR-w(jvmCompilerPipeline.kt:98)
        at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:30)
        at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:25)
        at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:19)
        at org.jetbrains.kotlin.cli.pipeline.PipelinePhase.phaseBody(PipelinePhase.kt:68)
        at org.jetbrains.kotlin.cli.pipeline.PipelinePhase.phaseBody(PipelinePhase.kt:58)
        at org.jetbrains.kotlin.config.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:102)
        at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:22)
        at org.jetbrains.kotlin.config.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:53)
        at org.jetbrains.kotlin.cli.pipeline.AbstractCliPipeline.runPhasedPipeline(AbstractCliPipeline.kt:109)
        at org.jetbrains.kotlin.cli.pipeline.AbstractCliPipeline.execute(AbstractCliPipeline.kt:68)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecutePhased(K2JVMCompiler.kt:51)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecutePhased(K2JVMCompiler.kt:42)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:90)
        at org.jetbrains.kotlin.cli.common.CLICompiler.exec(CLICompiler.kt:359)
        at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunnerBase.runCompiler(IncrementalJvmCompilerRunnerBase.kt:210)
        at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunnerBase.runCompiler(IncrementalJvmCompilerRunnerBase.kt:52)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.doCompile(IncrementalCompilerRunner.kt:509)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:421)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileNonIncrementally(IncrementalCompilerRunner.kt:303)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:130)
        at org.jetbrains.kotlin.buildtools.internal.jvm.operations.JvmCompilationOperationImpl.compileInProcess(JvmCompilationOperationImpl.kt:352)
        at org.jetbrains.kotlin.buildtools.internal.jvm.operations.JvmCompilationOperationImpl.compileInProcess(JvmCompilationOperationImpl.kt:282)
        at org.jetbrains.kotlin.buildtools.internal.jvm.operations.JvmCompilationOperationImpl.executeImpl(JvmCompilationOperationImpl.kt:102)
        at org.jetbrains.kotlin.buildtools.internal.jvm.operations.JvmCompilationOperationImpl.executeImpl(JvmCompilationOperationImpl.kt:65)
        at org.jetbrains.kotlin.buildtools.internal.CancellableBuildOperationImpl.execute(CancellableBuildOperationImpl.kt:58)
        at org.jetbrains.kotlin.buildtools.internal.KotlinToolchainsImpl$BuildSessionImpl.executeOperation$lambda$1(KotlinToolchainsImpl.kt:73)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.NoSuchMethodError: 'org.jetbrains.kotlin.ir.declarations.IrDeclarationOriginImpl org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin$Companion.getDEFINED()'
        at app.cash.burst.kotlin.IrKt.initDefaults(ir.kt:106)
        at app.cash.burst.kotlin.ClassSpecializer.createSubclass(ClassSpecializer.kt:147)
        at app.cash.burst.kotlin.ClassSpecializer.generateSpecializations(ClassSpecializer.kt:126)
        at app.cash.burst.kotlin.BurstIrGenerationExtension$generate$transformer$1.visitClassNew(BurstIrGenerationExtension.kt:55)
        at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitClass(IrElementTransformerVoidWithContext.kt:62)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:57)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:19)
        at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:72)
        at org.jetbrains.kotlin.ir.IrElementBase.transform(IrElementBase.kt:33)
        at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
        at org.jetbrains.kotlin.ir.declarations.IrPackageFragment.transformChildren(IrPackageFragment.kt:31)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitPackageFragment(IrElementTransformerVoid.kt:146)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:160)
        at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitFileNew(IrElementTransformerVoidWithContext.kt:122)
        at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitFile(IrElementTransformerVoidWithContext.kt:55)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:163)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:19)
        at org.jetbrains.kotlin.ir.declarations.IrFile.accept(IrFile.kt:27)
        at org.jetbrains.kotlin.ir.declarations.IrFile.transform(IrFile.kt:30)
        at org.jetbrains.kotlin.ir.declarations.IrFile.transform(IrFile.kt:19)
        at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
        at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.transformChildren(IrModuleFragment.kt:46)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:102)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:107)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:19)
        at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.accept(IrModuleFragment.kt:36)
        at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.transform(IrModuleFragment.kt:39)
        at app.cash.burst.kotlin.BurstIrGenerationExtension.generate(BurstIrGenerationExtension.kt:85)
        at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:578)
        ... 35 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,7 +3,9 @@
 ## [Unreleased]
 [Unreleased]: https://github.com/cashapp/burst/compare/2.11.0...HEAD
 
-Nothing yet!
+**Added**
+
+* Support Kotlin 2.3.20-Beta1
 
 
 ## [2.11.0] *(2025-12-16)*

--- a/burst-gradle-plugin/src/test/kotlin/app/cash/burst/gradle/klib.kt
+++ b/burst-gradle-plugin/src/test/kotlin/app/cash/burst/gradle/klib.kt
@@ -23,6 +23,7 @@ import kotlinx.metadata.klib.KlibModuleMetadata.MetadataLibraryProvider
 import org.jetbrains.kotlin.konan.file.File as KonanFile
 import org.jetbrains.kotlin.library.KotlinLibrary
 import org.jetbrains.kotlin.library.ToolingSingleFileKlibResolveStrategy
+import org.jetbrains.kotlin.library.components.metadata
 import org.jetbrains.kotlin.util.DummyLogger
 
 /** Decode a `.klib` file using APIs from the embeddable compiler. */
@@ -40,12 +41,12 @@ private fun KotlinLibrary.asMetadataLibraryProvider(): MetadataLibraryProvider {
   val original = this
   return object : MetadataLibraryProvider {
     override val moduleHeaderData: ByteArray
-      get() = original.moduleHeaderData
+      get() = original.metadata.moduleHeaderData
 
     override fun packageMetadata(fqName: String, partName: String) =
-      original.packageMetadata(fqName, partName)
+      original.metadata.getPackageFragment(fqName, partName)
 
     override fun packageMetadataParts(fqName: String): Set<String> =
-      original.packageMetadataParts(fqName)
+      original.metadata.getPackageFragmentNames(fqName)
   }
 }

--- a/burst-kotlin-plugin/src/main/kotlin/app/cash/burst/kotlin/BurstIrGenerationExtension.kt
+++ b/burst-kotlin-plugin/src/main/kotlin/app/cash/burst/kotlin/BurstIrGenerationExtension.kt
@@ -45,20 +45,6 @@ class BurstIrGenerationExtension(private val messageCollector: MessageCollector)
             return classDeclaration
           }
 
-          if (classHasAtBurst && classDeclaration.modality != Modality.ABSTRACT) {
-            try {
-              ClassSpecializer(
-                  pluginContext = pluginContext,
-                  burstApis = burstApis,
-                  originalParent = currentFile,
-                  original = classDeclaration,
-                )
-                .generateSpecializations()
-            } catch (e: BurstCompilationException) {
-              messageCollector.report(e.severity, e.message, currentFile.locationOf(e.element))
-            }
-          }
-
           // Snapshot the original functions because the loop mutates them.
           val originalFunctions = classDeclaration.functions.toList()
 
@@ -75,6 +61,20 @@ class BurstIrGenerationExtension(private val messageCollector: MessageCollector)
                   original = testFunction,
                 )
               specializer.generateSpecializations()
+            } catch (e: BurstCompilationException) {
+              messageCollector.report(e.severity, e.message, currentFile.locationOf(e.element))
+            }
+          }
+
+          if (classHasAtBurst && classDeclaration.modality != Modality.ABSTRACT) {
+            try {
+              ClassSpecializer(
+                  pluginContext = pluginContext,
+                  burstApis = burstApis,
+                  originalParent = currentFile,
+                  original = classDeclaration,
+                )
+                .generateSpecializations()
             } catch (e: BurstCompilationException) {
               messageCollector.report(e.severity, e.message, currentFile.locationOf(e.element))
             }

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -1,5 +1,5 @@
 [versions]
-kotlin = "2.3.0"
+kotlin = "2.3.20-Beta1"
 kotlinx-coroutines = "1.10.2"
 
 [libraries]

